src/lib/mod.rs

```diff
@@ -30,6 +30,40 @@
 // DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
+//! A library for manipulating Context Free Grammars (CFG). At the moment it only really supports
+//! Yacc-style grammars (albeit several variants of Yacc grammars), but the long-term aim is to
+//! provide an API that, when possible, is agnostic about the type of grammar being manipulated.
+//!
+//! A note on the terminology we use, since there's no universal standard (and EBNF, which is
+//! perhaps the closest we've got, uses terminology that now seems partially anachronistic):
+//!
+//!   * A rule is a mapping from a nonterminal name to 1 or more productions (the latter of which
+//!     is often called 'alternatives').
+//!   * A symbol is either a nonterminal or a terminal.
+//!   * A production is a (possibly empty) ordered sequence of symbols.
+//!
+//! Every nonterminal has a corresponding rule (and thus the two concepts are interchangeable);
+//! however, terminals are not required to appear in any production (such terminals can be used to
+//! catch error conditions).
```

Comment:
What do you mean by that? If no production has terminals then we'd have an empty language.

Reply:
You can define a terminal in a lexer that is never used in the grammar. IIRC Java uses this trick to reserve/prevent keywords.

Reply:
Ah that's how you meant this. The same is true for nonterminals as well though.

Reply:
Good point -- I hadn't thought of that. There is a difference though: all nonterminals (by definition) are defined in the grammar, though not all might be reachable from the start rule. [Should we classify such nonterms as an error or warning? Maybe...]

Reply:
Definitely not error as an error as that makes it harder to experiment with different rules without having to delete the non-used once. A warning might be helpful tho.

Reply:
Point taken. That's a job for the future though ;)